src/main/java/org/testng/TestRunner.java

```diff
@@ -1285,11 +1287,22 @@ public Injector getInjector(List<Module> moduleInstances) {
 
   @Override
   public Injector getInjector(IClass iClass) {
+    if (hasNoGuiceAnnotations(iClass)) {
```

Comment:
I think the fix won't work because the Injector definition won't exist if guice is not in the classpath.

Reply:
No. It works. I tested this out locally. `Guice` is our annotation. It won't need the Google dependency.

Reply:
@juherr - Also the problem is that when a user's test doesnt have anything to do with Guice DI framework, then `GuiceHelper` should not even be instantiated, because if it gets instantiated, then Java will end up attempting to load the `Module` class which will cause `NoClassDefFoundError`

Reply:
I agree with your explanation but how it is possible to load `com.google.inject.Injector`or 
`com.google.inject.Module` if guice.jar is missing (both are importing in the current class)?

Reply:
@juherr - Yes correct. This is the original error 

```
No application knows how to open /Users/703240805mac/temp/2405_project/target/surefire-reports/2020-11-23T16-08-48_974-jvmRun1.dump.
/Users/703240805mac/temp/2405_project  $ cat target/surefire-reports/2020-11-23T16-08-48_974-jvmRun1.dump
# Created at 2020-11-23T16:08:49.567
java.lang.BootstrapMethodError: java.lang.NoClassDefFoundError: com/google/inject/Module
	at org.testng.GuiceHelper.<clinit>(GuiceHelper.java:28)
	at org.testng.TestRunner.<init>(TestRunner.java:87)
	at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:596)
	at org.testng.SuiteRunner.init(SuiteRunner.java:173)
	at org.testng.SuiteRunner.<init>(SuiteRunner.java:107)
	at org.testng.TestNG.createSuiteRunner(TestNG.java:1300)
	at org.testng.TestNG.createSuiteRunners(TestNG.java:1276)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1125)
	at org.testng.TestNG.runSuites(TestNG.java:1063)
	at org.testng.TestNG.run(TestNG.java:1031)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:135)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:112)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:99)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.lang.NoClassDefFoundError: com/google/inject/Module
	... 18 more
Caused by: java.lang.ClassNotFoundException: com.google.inject.Module
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 18 more
```

Now after you defer the `GuiceHelper` instantiation to the first place where its used the new stacktrace shows up like this 

```
/Users/703240805mac/temp/2405_project  $ cat target/surefire-reports/2020-11-23T16-11-38_405-jvmRun1.dump
# Created at 2020-11-23T16:11:38.988
java.lang.BootstrapMethodError: java.lang.NoClassDefFoundError: com/google/inject/Module
	at org.testng.GuiceHelper.<clinit>(GuiceHelper.java:28)
	at org.testng.TestRunner.getInjector(TestRunner.java:1289)
	at org.testng.internal.ClassImpl.getInstanceFromGuice(ClassImpl.java:125)
	at org.testng.internal.ClassImpl.getDefaultInstance(ClassImpl.java:103)
	at org.testng.internal.ClassImpl.getInstances(ClassImpl.java:167)
	at org.testng.TestClass.getInstances(TestClass.java:102)
	at org.testng.TestClass.initTestClassesAndInstances(TestClass.java:82)
	at org.testng.TestClass.init(TestClass.java:74)
	at org.testng.TestClass.<init>(TestClass.java:39)
	at org.testng.TestRunner.initMethods(TestRunner.java:457)
	at org.testng.TestRunner.init(TestRunner.java:336)
	at org.testng.TestRunner.init(TestRunner.java:289)
	at org.testng.TestRunner.<init>(TestRunner.java:180)
	at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:596)
	at org.testng.SuiteRunner.init(SuiteRunner.java:173)
	at org.testng.SuiteRunner.<init>(SuiteRunner.java:107)
	at org.testng.TestNG.createSuiteRunner(TestNG.java:1300)
	at org.testng.TestNG.createSuiteRunners(TestNG.java:1276)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1125)
	at org.testng.TestNG.runSuites(TestNG.java:1063)
	at org.testng.TestNG.run(TestNG.java:1031)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:135)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:112)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:99)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.lang.NoClassDefFoundError: com/google/inject/Module
	... 29 more
Caused by: java.lang.ClassNotFoundException: com.google.inject.Module
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 29 more
```

Now when we fix the method `org.testng.TestRunner#getInjector(org.testng.IClass)` such that it first checks to see if the `@Guice` annotation was found before moving forward, basically solves the issue.

This is how the test project's dependency structure looks like

```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ 2405_project ---
[INFO] com.rationaleemotions:2405_project:jar:1.0-SNAPSHOT
[INFO] +- junit:junit:jar:4.11:test
[INFO] |  \- org.hamcrest:hamcrest-core:jar:1.3:test
[INFO] \- org.testng:testng:jar:7.4.0-SNAPSHOT:test
[INFO]    \- com.beust:jcommander:jar:1.78:test
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.038 s
[INFO] Finished at: 2020-11-23T16:14:25+05:30
[INFO] ------------------------------------------------------------------------
```

But I admit, I still dont have a proper explanation to your question, but what I am inferring here is that, as long as the value returned for the guice types are `null` I think Java is still fine with it and doesn't throw the `NoClassDefFoundError` necessarily.

Let me know if that is good enough an explanation to move forward with getting this changes merged

Reply:
The behavior is not what I expected but why not if it is working?

And what if the logic is inverted and `getInjector` never called if the class `hasNoGuiceAnnotations `?
https://github.com/cbeust/testng/blob/518bd75212805eb5f6f6a9518616abb5fd454159/src/main/java/org/testng/internal/ClassImpl.java#L103
and apply a similar pattern there https://github.com/cbeust/testng/blob/518bd75212805eb5f6f6a9518616abb5fd454159/src/main/java/org/testng/internal/Parameters.java#L646-L651

In fact, I don't know what is the expected behavior with `dataProviderClass` because they never have any `@Guice` annotation.

Reply:
@juherr - I am not sure why we need to be distributing the logic in 3 different places. I agree with the fact that I have not been able to explain as to how does Java not complain about that class during a null returning condition. But that's not a reason good enough to be splitting this logic in 3 places instead of putting them in one place (anyways all these code paths are converging at this point only right) ?

Reply:
@juherr - I think this SO post perhaps explains the behavior that we are right now seeing https://stackoverflow.com/a/45599139

Reply:
I updated my previous comment in order to highlight usages.

In fact, there are different subjects:
* I don't understand why the import is working when the jar is not present. 
  I think it is not a good practice to expose 3rd party API, even more, if it is an optional dependency.
  => LGTM: It works
* `hasNoGuiceAnnotations` doesn't have a lot of sense in the case of `Parameters`.
 => I feel it could not work in some corner cases with inheritance
* The different usages of `context.getInjector` are not very clear and the `Injector` could be easily hidden from the outside by something like:

```java
<T> T context.getInstance(Class<?> contextClass, Class<T> targetClass, Supplier<T> fallback) {
  Injector injector = getInjector(contextClass);
  if (injector == null) { 
    return fallback.get();  
  } 
  return injector.getInstance(targetClass);
}
``` 

Could you have a try?

Reply:
@juherr - Ok I didnt see this comment. My bad. 

> => I feel it could not work in some corner cases with inheritance

This is not a documented behaviour yet. So the interpretations are open I guess.

> The different usages of context.getInjector are not very clear and the Injector could be easily hidden from the outside by something like:

I think TestNG has already blended the Guice class references into `org.testng.ITestContext` via the below methods

```
List<Module> getGuiceModules(Class<? extends Module> cls);

default void addGuiceModule(Module module) {}

default List<Module> getAllGuiceModules() {
    return Collections.emptyList();
}

Injector getInjector(List<Module> moduleInstances);

Injector getInjector(IClass iClass);

void addInjector(List<Module> moduleInstances, Injector injector);
```

So its not very clear to me as to where you want me to add the suggested piece of code backed by a `Supplier`